docs/best-practices.rst

```diff
@@ -0,0 +1,296 @@
+In the case of :class:`Local executor <airflow.executors.local_executor.LocalExecutor>`, 
+storing a file on disk can make retries harder e.g., your task requires a config file that is deleted by another task in DAG.
+
+If possible, use ``XCom`` to communicate small messages between tasks and a good way of passing larger data between tasks is to use a remote storage such as S3/HDFS. 
+For example, if we have a task that stores processed data in S3 that task can push the S3 path for the output data in ``Xcom``,
+and the downstream tasks can pull the path from XCom and use it to read the data.
+
+The tasks should also not store any authentication parameters such as passwords or token inside them. 
+Where at all possible, use :ref:`Connections <concepts-connections>` to store data securely in Airflow backend and retrieve them using a unique connection id.
+
+
+Variables
+---------
+
+You should avoid usage of Variables outside an operator's ``execute()`` method or Jinja templates if possible, 
+as Variables create a connection to metadata DB of Airflow to fetch the value, which can slow down parsing and place extra load on the DB.
+
+Airflow parses all the DAGs in the background at a specific period.
+The default period is set using ``processor_poll_interval`` config, which is by default 1 second. During parsing, Airflow creates a new connection to the metadata DB for each DAG.
+It can result in a lot of open connections.
+
+The best way of using variables is via a Jinja template which will delay reading the value until the task execution. The template synaxt to do this is:
+
+.. code::
+
+    {{ var.value.<variable_name> }}
+
+or if you need to deserialize a json object from the variable :
+
+.. code::
+
+    {{ var.json.<variable_name> }}
+
+
+.. note::
+
+    In general, you should not write any code outside the tasks. The code outside the tasks runs every time Airflow parses the DAG, which happens every second by default.
+
+
+Testing a DAG
+^^^^^^^^^^^^^
+
+Airflow users should treat DAGs as production level code. The DAGs should have various tests to ensure that it produces expected results.
+You can write a wide variety of tests for a DAG. Let's take a look at some of them.
+
+DAG Loader Test
+---------------
+
+This test should ensure that your DAG does not contain a piece of code that raises error while loading.
+No additional code needs to be written by the user to run this test.
+
+.. code::
+
+ python your-dag-file.py
+
+Running the above command without any error ensures your DAG does not contain any uninstalled dependency, syntax errors, etc. 
+
+You can look into :ref:`Testing a DAG <testing>` for details on how to test individual operators.
+
+Unit tests
+-----------
+
+Unit tests ensure that there is no incorrect code in your DAG. You can write a unit test for your tasks as well as your DAG.
+
+**Unit test for loading a DAG:**
+
+.. code::
```

Comment:
```suggestion
.. code:: python
```

Reply:
(FYI this one isn't required it seems as python is configured as the default lanauge. I think)

Reply:
Oh yes I think so too. I saw at a different doc we don't mention it but it still renders it as python.

https://github.com/apache/airflow/pull/6515#discussion_r350229921